Lock ideas once they're closed and lock voting stats

ThatTobMate opened the pull request:
This is a bit of a spike. I saw a community member mention that one of their old proplot submissions has lost a bunch of votes. With the current way we do things this is possible in the following circumstance:

- User 1 votes on idea 1 with 100 tokens
- Idea 1 has 101 upvotes including the 1 vote from the Idea proposer
- User 1 logs out and later sells 99 tokens
- User 1 revisits PropLot, logs back in. We update User 1 lilnounCount from 100 to 1.
- Idea 1 will now lose 99 votes as we calculate the votes dynamically on each request by looking at the voters current lilnounCount.

I think this is fine in normal circumstances while the vote is still open, however once the voting has closed we should lock those results in place. Especially as we want to build a profile of user/idea voting history.

This PR is an attempt at locking ideas and the voting statistics in place once we classify them as closed. The way I have this set up now is that when a user loads the proplot list we asynchronously update the ideas that have now "closed" with the voting statistics and no longer dynamically calculate the votes for these ideas on future requests.

I think this should work but it does mean we need regular traffic to ensure closed ideas are regularly locked. We could also implement a cron job that does this on a daily basis. I think this logic should be fine though for now?

Any thoughts on this issue + approach @mcgingras ?

ThatTobMate commented on packages/nouns-api/src/services/ideas.ts:
```diff
@@ -96,15 +74,31 @@ class IdeasService {
+
+            if (closed) {
+              //Trigger async event to update idea
+              ideasToUpdate.push({ ...idea, votecount, closed, upvotes, downvotes });
```
Feels a bit dirty to be running these updates on the request to fetch the list but it should happen async to the incoming request and infrequently enough that it shouldn't be an issue or cause additional latency.

mcgingras commented on packages/nouns-api/src/services/ideas.ts:
```diff
@@ -379,7 +396,23 @@ class IdeasService {
+      await prisma.idea.update({
+        where: { id: idea.id },
+        data: {
+          locked: true,
```
why do we have to reset this lock to true? I thought these were already locked?
This is what we are doing above -- 
I'm confused how they are originally getting their locked status

```
let closed = idea.locked;
if (!closed) { // add to list that gets sent to batchLockIdeas }
```

ThatTobMate commented on packages/nouns-api/src/services/user.ts:
```diff
@@ -140,9 +106,16 @@ class UserService {
         if (comment.idea) {
-          const votecount = calculateVotes(comment.idea.votes || []);
+          let votecount = comment.idea.netVotes || 0;
+          let closed = comment.idea.locked;
+          if (!closed) {
+            const { count } = calculateAllVotes(comment.idea.votes || []);
+            votecount = count;
+            closed = getIsClosed(comment.idea);
```
I think we cleared this up in Discord but "closed" previously has been a dynamically applied property to ideas based on the createdAt date being > 7 days. The `locked` property is a new property added to the idea in the db.

We still want to check for the "closed" tag if the `locked` property is false as the locked value will only get set when the cron job runs, there may be a period of time where an idea is closed but the cron hasn't yet updated the locked property.

ThatTobMate commented:
@mcgingras I've updated this to now use a cron job, i've set this to run as a seperate process in our fly deploy. There will now be a task runing the web app and one running the cron. It's set to run once and hour at the moment.

I also updated the migrate command that we've been runnin on production builds to `prisma migrate deploy`, we were running `prisma migrate dev` 😅
